Enhance frictional models based on ExternalCombiTable1D

beutlich opened the pull request:
@christoff-buerger This adds your proposal as PR. Note that I swapped the components of table_signs to match the columns of mu_pos.

Since CombiTable1Ds supports more interpolation smoothness kinds we also could get rid of

https://github.com/modelica/ModelicaStandardLibrary/blob/81dd1b63a61075c85fbd26945eb333b202b9e4b9/Modelica/Mechanics/Rotational/Components/Clutch.mo#L102-L103

, but might lose mu0 being a parameter again.

Closes #3661

tobolar commented on Modelica/Mechanics/Rotational/Components/Clutch2.mo:
```diff
@@ -0,0 +1,171 @@
+  tau0_max = peak*tau0;
+
+  // Friction torque
+  table_signs = if startForward then
+      {1,1}
+    elseif startBackward then
+      {1,-1}
+    elseif pre(mode) == Forward then
+      {1,1}
+    else
+      {-1,-1};
```
Possible to change to (for even better readability):
```
  table_signs =
    if     startForward         then { 1, 1}
    elseif startBackward        then { 1,-1}
    elseif pre(mode) == Forward then { 1, 1}
    else                             {-1,-1};
```

beutlich replied:
Sure, go ahead.

beutlich commented on Modelica/Mechanics/Rotational/Components/Clutch2.mo:
```diff
@@ -0,0 +1,168 @@
+    else                             {-1,-1};
+  table.u = table_signs[1]*w_rel;
+  mu = table_signs[2]*table.y[1];
+  tau = if locked then sa*unitTorque elseif free then 0 else cgeo*fn*mu;
```
```suggestion
  tau = if locked then sa*unitTorque elseif free then 0 else mu*cgeo*fn;
```

Have it similar as tau0.

otronarp commented:
May a suggest a name change? Calling things `Foo2` is not helpful to the user unless the two signifies that there are two of something within the model (and as far as I understand it that's not the case here, it's an alternative implementation right?).

beutlich commented:
> 
> 
> May I suggest a name change? Calling things `Foo2` is not helpful to the user unless the two signifies that there are two of something within the model (and as far as I understand it that's not the case here, it's an alternative implementation right?).

I only did so for the sake of consistency since we alraedy have ElastoBacklash and ElastoBacklash2 in Rotational.Components (and even other component models like Diode and Diode2).

tobolar commented:
> It is an slightly improved implementation.

Actually, this reimplementation shall have the same functionality as the original clutch model. So why is it necessary to double the it?

tobolar commented:
We shall also reimplement `Mechanics.Rotational.Components.Brake` accordingly.

MartinOtter reviewed (changes requested):
Very good proposal (the "current" friction elements had been implemented at a time, where no table interpolation was available and therefore a quick implementation with a Modelica function was made;).

However, it would be good to use the new friction elements at least in one example (maybe I missed this).

beutlich commented:
> 
> However, it would be good to use the new friction elements at least in one example (maybe I missed this).

I already updated Modelica.Mechanics.Rotational.Examples.CoupledClutches to utilize Clutch2.

> Currently, only linear interpolation in the table is supported.

@MartinOtter Do you agree to also add the interpolation smoothness as new parameter? In that case we no longer should use Modelica.Math.Vectors.interpolate to determine mu0.

MartinOtter commented:
> 
> 
> > However, it would be good to use the new friction elements at least in one example (maybe I missed this).
> 
> I already updated Modelica.Mechanics.Rotational.Examples.CoupledClutches to utilize Clutch2.

O.k.

> 
> > Currently, only linear interpolation in the table is supported.
> 
> @MartinOtter Do you agree to also add the interpolation smoothness as new parameter? In that case we no longer should use Modelica.Math.Vectors.interpolate to determine mu0.

Yes, smoothness makes case and mu0 should no longer be used with Vectors.interpolate.

beutlich commented:
> > @MartinOtter Do you agree to also add the interpolation smoothness as new parameter? In that case we no longer should use Modelica.Math.Vectors.interpolate to determine mu0.
> 
> Yes, smoothness makes sense and mu0 should no longer be used with Vectors.interpolate.

Done for Clutch2 for first review. @tobolar If you agree we should migrate Brake2 and OneWayClutch2 from block CombiTable1D to external object ExternalCombiTable1D, too.

beutlich commented:
> Actually, this reimplementation shall have the same functionality as the original clutch model. So why is it necessary to double the it?

I agree that the revised models Clutch2/Brake2/OneWayClutch2 fall back to the same functionality as the legacy models Clutch/Brake/OneWayClutch. There is no particular reason to not simply update the original models. @MartinOtter needs to decide here.

christoff-buerger commented:
> Note that I swapped the components of table_signs to match the columns of mu_pos.

That is good (= more intuitive).

 > May I suggest a name change? Calling things Foo2 is not helpful to the user unless the two signifies that there are two of something within the model (and as far as I understand it that's not the case here, it's an alternative implementation right?).

 > It is an slightly improved implementation. I only did so for the sake of consistency since we already have ElastoBacklash and ElastoBacklash2 in Rotational.Components (and even other component models like Diode and Diode2).

I agree with @otronarp. We should avoid doubling and names like `Foo2`.

Is there any rationale or argument why we need two clutch implementations; why we have to keep the old as well? Does the redesign have any drawbacks (functional or non-functional like useabelility or performance)?

This is particularly important because

 > We shall also reimplement Mechanics.Rotational.Components.Brake and Mechanics.Rotational.Components.OneWayClutch accordingly.

which would result in a plethora of `Foo2` things.

Do we also double the test models/examples utilizing the redesign? Or would we just not use one of the two implementations?

Thus, I like to push on:

 > I agree that the revised models Clutch2/Brake2/OneWayClutch2 fall back to the same functionality as the legacy models Clutch/Brake/OneWayClutch. There is no particular reason (except that protected mu0 changed variability from variable to parameter) to not simply update the original models. @MartinOtter needs to decide here. If he agrees we can avoid the duplication of the models.

@MartinOtter what is your opinion?

christoff-buerger commented:
@tobolar: The `Brake2` still uses:
```
 final parameter Real mu0 = Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], 0, 1)
```
i.e., always linear segments as smoothness; likewise the `OneWayClutch2`.

I propose to handle it like @beutlich improved for `Clutch2`, i.e., to support different table smoothness.

christoff-buerger commented:
Above two things, (1) can we avoid doubling of clutch, brake etc models and (2) consistent support for full table smoothness, have to be addressed from my point of view before a positive review.

beutlich commented:
> 
> 
> Above two things, (1) can we avoid doubling of clutch, brake etc models and (2) consistent support for full table smoothness, have to be addressed from my point of view before a positive review.

Note, that I added Clutch2 and updated it with full table smoothnes just as an offer to discuss with the library officers to get their opinion which way to head. Once this is clarifief we can modify/add Brake and OneWayClutch accordingly.

tobolar commented:
`Clutch2` looks fine to me now and the smoothness option is great.

Since the piece of code 
```
    if     smoothness == Types.Smoothness.ConstantSegments then TabInternal.getTable1DValueNoDer(tableID, 1, 0)
    elseif smoothness == Types.Smoothness.LinearSegments   then TabInternal.getTable1DValueNoDer2(tableID, 1, 0)
    else                                                        TabInternal.getTable1DValue(tableID, 1, 0)
```
will be used also in other elements (and they are not only Rotational, see #3664), we shall extract it into a separate function. What could be the best location for such a function?

beutlich commented on Modelica/Mechanics/Rotational/Components/Clutch2.mo:
```diff
@@ -3,9 +3,12 @@ model Clutch2 "Clutch based on Coulomb friction (with interpolation by ExternalC
   extends Modelica.Mechanics.Rotational.Interfaces.PartialCompliantWithRelativeStates;
 
+  import Modelica.Blocks.Types;
+  import TabInternal = Modelica.Blocks.Tables.Internal;
```
Renaming imports should be avoided. Better:

```mo
import Modelica.Blocks.Tables;
```

beutlich commented:
> getTable1DValue

I also thought about this. My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) and calling the functions from Internal. @tobolar Can you do it?

tobolar commented:
> Can you do it?

I will prepare a suggestion.

beutlich commented on Modelica/Mechanics/Rotational/Components/Clutch2.mo:
```diff
@@ -7,11 +7,11 @@ model Clutch2 "Clutch based on Coulomb friction (with interpolation by ExternalC
 
   parameter Real mu_pos[:, 2] = [0, 0.5]
-    "Positive sliding friction coefficient [-] as function of w_rel [rad/s] (w_rel>=0)";
+    "Positive sliding friction coefficient [-] as function of w_rel [rad/s] (w_rel >= 0)";
```
I'd actually avoid to introduce cosmetical changes now. This should go to a separate issue addressing all component models.

tobolar replied:
Ok, how to revert this?

beutlich replied:
> 
> 
> Ok, how to revert this?

```
git reset ca8078fbd49f1f9ab892818bbcd93a3d60ac12f1
git push --force
```

tobolar replied:
Done. The first time I've ever did it!

beutlich replied:
You still have the changes locally. So be careful when pushing again.

tobolar replied:
I use TortoiseSVN which enables three options how to reset a commit. I selected the one where also local changes are reset.
But anyway, thanks for the hint because I've never done it so far.

dietmarw replied:
I truly hope you are **not** using Tprotois**SVN** in order to work on the git repo ;-)

tobolar commented:
@beutlich 
> My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) ...

Actually, the smoothness is already given in the constructor of ExternalCombiTable1D. Is it reasonable to have it as an additional argument then?

beutlich commented:
> 
> 
> @beutlich
> 
> > My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) ...
> 
> Actually, the smoothness is already given in the constructor of ExternalCombiTable1D. Is it reasonable to have it as an additional argument then?

Yes, since we decide in the Modelica side which if the functions to call.

HansOlsson commented:
> > What could be the best location for such a function?
> 
> I also thought about this. My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) and calling the functions from Internal. @tobolar Can you do it?

Note that inlining of this function is very important, as we want to be able to differentiate for some values of smoothness, but cannot for others. If we cannot guarantee that inline+optional differentiation doesn't work we shouldn't introduce the new function.

tobolar commented:
Trying to implement the function. But having smoothness as an argument can lead to undesired results, see example below where parameters `smoothnessConstructor` and `smoothnessEvaluator` have different values. Or do I understand it wrong?

```
function getTable1DValue "Interpolate one-dimensional table defined by matrix"
  import Modelica.Blocks.Types.Smoothness;
  import Modelica.Blocks.Tables.Internal;

  input Modelica.Blocks.Types.ExternalCombiTable1D tableID;
  input Integer icol;
  input Smoothness smoothness = Smoothness.LinearSegments;
  input Real u;
  output Real y;
algorithm 
  y :=
    if     smoothness == Smoothness.ConstantSegments then Internal.getTable1DValueNoDer(tableID, icol, u)
    elseif smoothness == Smoothness.LinearSegments   then Internal.getTable1DValueNoDer2(tableID, icol, u)
    else                                                  Internal.getTable1DValue(tableID, icol, u);
end getTable1DValue;

model Test
  import Modelica.Blocks.Types;

  parameter Real table[:, 2] = [0, 0.5;1, 1; 2,3; 3,4; 4,2; 5,2.5; 6,3];
  parameter Types.Smoothness smoothnessConstructor = Types.Smoothness.LinearSegments;
  parameter Types.Smoothness smoothnessEvaluator = Types.Smoothness.ConstantSegments;
  Real y;
protected 
  final parameter Types.ExternalCombiTable1D tableID = Types.ExternalCombiTable1D(
    tableName="NoName",
    fileName="NoName",
    table=table,
    columns={2},
    smoothness=smoothnessConstructor,
    extrapolation=Types.Extrapolation.LastTwoPoints,
    verboseRead=false);
equation 
  y = getTable1DValue(tableID, 1, smoothnessEvaluator, time);
end Test;
```

beutlich commented:
> Or do I understand it wrong?

I see your point even though it is not relevant for the current use case. Unfortunately we do not (yet) have a getter function of the external object to retrieve its ctor settings. To keep things simple I propose to leave the table value call as is, even if it means to write duplicate code and long lines. (It is in a protected section anyway.)

beutlich commented:
> Above two things, (1) can we avoid doubling of clutch, brake etc models and (2) consistent support for full table smoothness, have to be addressed from my point of view before a positive review.

I refactored again and now there are no more duplicated classes.

beutlich commented:
What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.

beutlich commented:
Please also not note that chaning mu0 from public to protected section is a backward-compatibility changing break that needs at least be documented (in the release notes).

tobolar commented:
> Please also not note that chaning mu0 from public to protected section is a backward-compatibility breaking change...

We can also make it public again and protect it in the next major version. But actually, I can hardly imagine mu0 is a matter of interesse of a user.

tobolar reviewed (approved):
Fine for me now.

tobolar reviewed (changes requested):
BearingFriction and LossyGear seem to be affected as well. I can care.

tobolar commented:
Now BearingFriction and LossyGear are changed as well.

In LossyGear, variable `ideal` can be changed to parameter as well.

beutlich commented:
> In LossyGear, variable `ideal` can be changed to parameter as well.

`ideal` is an unnecessary optimization that can be removed. Also `interpolation_result` is not required at all, especially not as matrix.

tobolar commented:
I agree. What about the backward compatibility.

HansOlsson commented on Modelica/Mechanics/Rotational/Components/LossyGear.mo:
```diff
@@ -26,7 +26,7 @@ model LossyGear
     "Angular acceleration of flange_a with respect to support";
 
-  Real interpolation_result[1, 4]
+  Real interpolation_result[1, nCol-1]
```
It is not clear to me that using a constant is better here. You now see `nCol-1` and also a description that describes **four** columns; so I'm more left wondering whether they are the same or not. As far as I understand there is no plan to make nCol variable. If we are going with such a constant it would to me be more logical to have nValueCols=4, and replace nCol and nCol-1 with nValueCols+1 and nValueCols; as it seems more normal to add columns than to remove them.

tobolar replied:
The introduction of nCol is rather intended for consistency in dimension of lossTable and constructor of tableID. But it shall also be consistent for interpolation_result. So I simply took a short named constant for it.

I have no preference for it - we can also call it e.g. nValue=4.

beutlich replied:
I propose to remove constant nCol again. It is protected, but SimulationX displays it as dimensions of lossTable where nCol cannot be deduced at all.

tobolar replied:
Ok, I will revert the changes on it.

tobolar replied:
Reverted with 8a2bb0a7a3d336c5b49537e3de01feef984d3b6d

tobolar reviewed (approved):
Fine for me.

beutlich commented:
> 
> 
> What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.

@HansOlsson @christoff-buerger Can you please confirm that the Evaluate annotation for mu0 should be removed again. Thanks.

HansOlsson commented:
> > What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.
> 
> @HansOlsson @christoff-buerger Can you please confirm that the Evaluate annotation for mu0 should be removed again. Thanks.

To me "Evaluate=true" should be removed as mu0 should not be evaluated during translation, since it depends on the external table (and evaluating that during translation requires quite specific solution and doesn't really help).

HansOlsson reviewed (changes requested):
As indicated remove the Evaluate=true for mu0.

beutlich commented:
> 
> 
> As indicated remove the Evaluate=true for mu0.

Done. Please update your review. Thank you again!

HansOlsson reviewed (approved):
Looks good now.

beutlich commented:
OK, waiting for @christoff-buerger's final approval then.

MartinOtter commented:
>    I agree that the revised models Clutch2/Brake2/OneWayClutch2 fall back to the same functionality as the legacy models Clutch/Brake/OneWayClutch. There is no particular reason (except that protected mu0 changed variability from variable to parameter) to not simply update the original models. @MartinOtter needs to decide here. If he agrees we can avoid the duplication of the models.

As long as the changes are backwards compatible (e.g. only new parameters added, but not old parameters removed or changed), it is of course better to have only one model and not two different models.

christoff-buerger commented:
> 
> 
> > > What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.
> > 
> > 
> > @HansOlsson @christoff-buerger Can you please confirm that the Evaluate annotation for mu0 should be removed again. Thanks.
> 
> To me "Evaluate=true" should be removed as mu0 should not be evaluated during translation, since it depends on the external table (and evaluating that during translation requires quite specific solution and doesn't really help).

The "Evaluate=true" stems from the original redesign where mu0 was still defined by means of `Modelica.Math.Vectors.interpolate` and not a combi-table.

Since we decided since than that full table interpolation modes should be supported (not just linear), we also have to define mu0 via a combi-table now (because we need to support spline interpolation & co).

From a design point, mu0 is a constant. It is pity that we cannot use "Evaluate=true" with it because combi-tables as external objects are challenging for Modelica tools. To that end, I see your concern as a tool shortcoming/challenge.

It is an important point however -- that tables are not suited for "Evaluate=true" -- such that this annotation should be dropped.

christoff-buerger reviewed (approved):
Looks good. We have all relevant classes redesigned to use tables instead of `Modelica.Math.Vectors.interpolate` and support all kind of table smoothness/interpolation modes.

There are no class / functionality duplications like `Clutch2` etc.

We also have no `Evaluate=true` on constants derived via table interpolation, since such is troublesome for external table objects (a _common_ tooling shortcoming).

christoff-buerger commented:
It can be merged from my perspective.
